TESTS/mbedmicro-rtos-mbed/signals/main.cpp

```diff
@@ -15,49 +15,357 @@
+
+    Given a thread B running
+    When thread B executes @a signal_wait(0, 0) with no signal specified for 0[ms] timeout
+    Then thread B @a signal_wait status should be osOK indicating correct exit
+ */
+void test_wait_0(void)
+{
+    Thread tb(osPriorityNormal, TEST_STACK_SIZE);
+    tb.start(callback(run_signal_wait<0, 0>));
+
+    sem_a.wait();
+    TEST_ASSERT_EQUAL(osOK, ret_status);
+}
+
+/** Test signal_wait timeout
+
+    Given a thread B running
+    When thread B executes @a signal_wait(signals, timeout) with specified signal and timeout
+    Then thread B @a signal_wait status should be osEventTimeout indicating a timeout
+ */
+template <int32_t signals, uint32_t timeout>
+void test_wait_timeout(void)
+{
+    Thread tb(osPriorityNormal, TEST_STACK_SIZE);
+    tb.start(callback(run_signal_wait<signals, timeout>));
+
+    sem_a.wait();
+    TEST_ASSERT_EQUAL(osEventTimeout, ret_status);
+}
+
+/** Test signal_wait on thread with all signals already set
+
+    Given two threads A and B running
+    When thread A executes @a signal_set(ALL_SIGNALS) with all possible signals
+    Then thread B executes @a signal_wait(ALL_SIGNALS, osWaitForever), @a signal_wait return immediately with
+        status osEventSignal indicating a all wait signals was already set
+ */
+void test_wait_all_already_set(void)
+{
+    int32_t ret;
+    Thread tb(osPriorityNormal, TEST_STACK_SIZE);
+    tb.start(callback(run_signal_wait3<ALL_SIGNALS, osWaitForever>));
+
+    sem_a.wait();
+    ret = tb.signal_set(ALL_SIGNALS);
+    TEST_ASSERT_EQUAL(ALL_SIGNALS, ret);
+    sem_b.release();
 
-    Thread thread(osPriorityNormal, TEST_STACK_SIZE);
-    thread.start(led_thread);
+    sem_a.wait();
+    TEST_ASSERT_EQUAL(osEventSignal, ret_status);
+}
 
-    bool result = false;
+/** Test signal_wait on thread with no signals already set
 
-    printf("Handling %d signals...\r\n", SIGNALS_TO_EMIT);
+    Given two threads A and B running and B waiting for a thread flag to be set
+    When thread A executes @a signal_set(ALL_SIGNALS) with all possible signals
+    Then thread B @a signal_wait status is osEventSignal indicating a all wait signals was set
+ */
+void test_wait_all(void)
+{
+    int32_t ret;
+    Thread tb(osPriorityNormal, TEST_STACK_SIZE);
+    tb.start(callback(run_signal_wait2<ALL_SIGNALS, osWaitForever>));
+
+    sem_a.wait();
+    TEST_ASSERT_EQUAL(Thread::WaitingThreadFlag, tb.get_state());
+
+    ret = tb.signal_set(ALL_SIGNALS);
+    TEST_ASSERT_EQUAL(0, ret);
+
+    sem_a.wait();
+    TEST_ASSERT_EQUAL(osEventSignal, ret_status);
+}
 
-    while (true) {
-        Thread::wait(2 * SIGNAL_HANDLE_DELEY);
-        thread.signal_set(SIGNAL_SET_VALUE);
-        if (signal_counter == SIGNALS_TO_EMIT) {
-            printf("Handled %d signals\r\n", signal_counter);
-            result = true;
-            break;
-        }
+/** Test signal_wait on thread with no signals already set - signals set one by one in loop
```

Comment:
Could the first line for all test document, on a high level, what the tests asserts? Something like "Validate that set signals accumulate and cause signal wait to return"